File: erezervisi_api/eRezervisi/eRezervisi.Common.Dtos/AccommodationUnit/AccommodationUnitGetDto.cs
﻿using eRezervisi.Common.Dtos.AccommodationUnitCategories;
using eRezervisi.Common.Dtos.AccommodationUnitPolicy;
using eRezervisi.Common.Dtos.Image;
using eRezervisi.Common.Dtos.Township;
using eRezervisi.Core.Domain.Enums;

namespace eRezervisi.Common.Dtos.AccommodationUnit
{
    public class AccommodationUnitGetDto
    {
        public long Id { get; set; }
        public string Title { get; set; } = null!;
        public string Address { get; set; } = null!;
        public double Price { get; set; }
        public string? Note { get; set; }
        public long OwnerId { get; set; }
        public PolicyGetDto AccommodationUnitPolicy { get; set; } = null!;
        public long AccommodationUnitCategoryId { get; set; }
        public CategoryGetDto AccommodationUnitCategory { get; set; } = null!;
        public long TownshipId { get; set; }
        public TownshipGetDto Township { get; set; } = null!;
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public double TapPosition { get; set; }
        public string ThumbnailImage { get; set; } = null!;
        public List<ImageGetDto> Images { get; set; } = new();
        public bool Favorite { get; set; }
        public double? AverageRating { get; set; }
        public AccommodationUnitStatus Status { get; set; }
    }
}

File: erezervisi_mobile/erezervisi_mobile/lib/models/responses/accommodation_unit/accommodation_unit_get_dto.dart
import 'package:erezervisi_mobile/enums/accommodation_unit_status.dart';
import 'package:erezervisi_mobile/models/responses/category/category_get_dto.dart';
import 'package:erezervisi_mobile/models/responses/image/image_get_dto.dart';
import 'package:erezervisi_mobile/models/responses/policy/policy_get_dto.dart';
import 'package:erezervisi_mobile/models/responses/township/township_get_dto.dart';

class AccommodationUnitGetDto {
  late num id;
  late String title;
  late num price;
  late String? note;
  late PolicyGetDto? policy;
  late CategoryGetDto? category;
  late TownshipGetDto township;
  late num latitude;
  late num longitude;
  late String thumbnailImage;
  late List<ImageGetDto> images;
  late bool favorite;
  late num ownerId;
  late String address;
  late AccommodationUnitStatus status;

  AccommodationUnitGetDto(
      {required this.id,
      required this.title,
      required this.price,
      this.note,
      required this.policy,
      required this.category,
      required this.township,
      required this.latitude,
      required this.longitude,
      required this.thumbnailImage,
      required this.images,
      required this.favorite,
      required this.ownerId,
      required this.address,
      required this.status});

  factory AccommodationUnitGetDto.fromJson(Map<String, dynamic> json) {
    return AccommodationUnitGetDto(
        id: json['id'] as num,
        title: json['title'],
        price: json['price'] as num,
        policy: json['accommodationUnitPolicy'] != null
            ? PolicyGetDto.fromJson(json['accommodationUnitPolicy'])
            : null,
        category: json['category'] != null
            ? CategoryGetDto.fromJson(json['accommodationUnitCategory'])
            : null,
        township: TownshipGetDto.fromJson(json['township']),
        latitude: json['latitude'] as num,
        longitude: json['longitude'] as num,
        thumbnailImage: json['thumbnailImage'],
        favorite: json['favorite'],
        ownerId: json['ownerId'] as num,
        address: json['address'],
        images: (json['images'] as List<dynamic>)
            .map((imageJson) => ImageGetDto.fromJson(imageJson))
            .toList(),
        status: AccommodationUnitStatus.values
            .where((item) => item.index == json['status'])
            .first);
  }
}
